src/lib/optimizer/optimizer.cpp

```diff
@@ -172,11 +174,31 @@ std::shared_ptr<AbstractLQPNode> Optimizer::optimize(std::shared_ptr<AbstractLQP
   const auto root_node = LogicalPlanRootNode::make(std::move(input));
   input = nullptr;
 
-  if constexpr (HYRISE_DEBUG) validate_lqp(root_node);
+  if constexpr (HYRISE_DEBUG) {
```

Comment:
Could you explain what you added here?

Reply:
i thought we don't want the timer and rule name deduction in there for release builds. before, we had the validation steps before and after the rule execution in debug builds. now, it's either clean rule execution or validation, timed rule execution, validation.

Reply:
I'd say that release builds are much more interesting. Maybe you could show a comparison of the breakdown for debug and release builds? I'd expect there to be some bigger differences.

Reply:
Agreed! Performance measurements for debug builds have usually a low significance. But on the other hand, it would be interesting to see if this timing has any measurable impact on benchmark performance.

Reply:
I compared the TPCC benchmarks in release build with and without optimizer rule timing and it seems that it has indeed an impact. 5% seem a little bit too much for enabeling it by default. what do you think about another benchmark configuration that enables the timing only if needed?

```
+----------------+----------------+------+----------------+------+------------+----------------------+
| Benchmark      | prev. iter/s   | runs | new iter/s     | runs | change [%] |              p-value |
+----------------+----------------+------+----------------+------+------------+----------------------+
| Delivery       | 0.911102354527 | 164  | 0.866361558437 | 156  | -5%        | (run time too short) |
| New-Order      | 10.2332353592  | 1842 | 9.7576751709   | 1757 | -5%        |               0.0000 |
| Order-Status   | 0.90554690361  | 163  | 0.866361558437 | 156  | -4%        | (run time too short) |
| Payment        | 9.78879547119  | 1762 | 9.32449436188  | 1679 | -5%        | (run time too short) |
| Stock-Level    | 0.90554690361  | 163  | 0.866361558437 | 156  | -4%        | (run time too short) |
| geometric mean |                |      |                |      | -5%        |                      |
+----------------+----------------+------+----------------+------+------------+----------------------+

```

Reply:
`--metrics`?

Reply:
`--optimizer-metrics`?

Reply:
Do we want/need different metrics levels?

Reply:
Result of the fortnightly: Let's check if we can reduce the tracking cost first.

Reply:
it was indeed the string demangling that added so much time. if we do that in the python script, we get only -1% performance overhead. could we live with that?

```
+----------------+----------------+------+----------------+------+------------+----------------------+
| Benchmark      | prev. iter/s   | runs | new iter/s     | runs | change [%] |              p-value |
+----------------+----------------+------+----------------+------+------------+----------------------+
| Delivery       | 0.911095440388 | 164  | 0.899798333645 | 162  | -1%        | (run time too short) |
| New-Order      | 10.2220458984  | 1840 | 10.1588344574  | 1829 | -1%        |               0.6388 |
| Order-Status   | 0.905539989471 | 163  | 0.899798333645 | 162  | -1%        | (run time too short) |
| Payment        | 9.76094341278  | 1757 | 9.71448898315  | 1749 | -0%        | (run time too short) |
| Stock-Level    | 0.911095440388 | 164  | 0.899798333645 | 162  | -1%        | (run time too short) |
| geometric mean |                |      |                |      | -1%        |                      |
+----------------+----------------+------+----------------+------+------------+----------------------+

```

Reply:
Could you run TPC-H with SF 0.01 just to get a second data point, please?

Reply:
For TPC-H also -1% mean:

```
+----------------+---------------+--------+---------------+--------+------------+---------+
| Benchmark      | prev. iter/s  | runs   | new iter/s    | runs   | change [%] | p-value |
+----------------+---------------+--------+---------------+--------+------------+---------+
| TPC-H 01       | 69.6706771851 | 4181   | 66.8618774414 | 4012   | -4%        |  0.0000 |
| TPC-H 02       | 517.577331543 | 31055  | 520.63079834  | 31238  | +1%        |  0.4273 |
| TPC-H 03       | 593.663513184 | 35620  | 595.178283691 | 35711  | +0%        |  0.0072 |
| TPC-H 04       | 886.466186523 | 53188  | 888.989257812 | 53340  | +0%        |  0.0000 |
| TPC-H 05       | 355.508605957 | 21331  | 356.893920898 | 21414  | +0%        |  0.1542 |
| TPC-H 06       | 3368.53833008 | 202113 | 3361.73608398 | 201705 | -0%        |  0.0000 |
| TPC-H 07       | 236.379425049 | 14183  | 230.260604858 | 13816  | -3%        |  0.1846 |
| TPC-H 08       | 24.7375068665 | 1485   | 24.7477493286 | 1485   | +0%        |  0.9699 |
| TPC-H 09       | 190.273300171 | 11417  | 189.314483643 | 11359  | -1%        |  0.3475 |
| TPC-H 10       | 324.644439697 | 19479  | 323.588653564 | 19416  | -0%        |  0.0000 |
| TPC-H 11       | 1094.73828125 | 65685  | 1099.55517578 | 65974  | +0%        |  0.0000 |
| TPC-H 12       | 726.192687988 | 43572  | 728.618591309 | 43718  | +0%        |  0.0010 |
| TPC-H 13       | 256.077880859 | 15365  | 240.425704956 | 14426  | -6%        |  0.0000 |
| TPC-H 14       | 1389.7442627  | 83385  | 1394.77392578 | 83687  | +0%        |  0.0000 |
| TPC-H 15       | 391.153015137 | 23470  | 373.581420898 | 22415  | -4%        |  0.0000 |
| TPC-H 16       | 241.167785645 | 14471  | 236.514419556 | 14191  | -2%        |  0.0000 |
| TPC-H 17       | 1198.90344238 | 71935  | 1203.57421875 | 72215  | +0%        |  0.0504 |
| TPC-H 18       | 200.514389038 | 12031  | 200.445114136 | 12027  | -0%        |  0.3096 |
| TPC-H 19       | 99.2068862915 | 5953   | 98.2062072754 | 5893   | -1%        |  0.0000 |
| TPC-H 20       | 189.089447021 | 11346  | 186.923522949 | 11216  | -1%        |  0.0000 |
| TPC-H 21       | 104.742706299 | 6285   | 104.421081543 | 6266   | -0%        |  0.3499 |
| TPC-H 22       | 340.831390381 | 20450  | 336.041107178 | 20163  | -1%        |  0.0000 |
| geometric mean |               |        |               |        | -1%        |         |
+----------------+---------------+--------+---------------+--------+------------+---------+

```